Added all the autogenerated files.

alex4o opened the pull request:
This is the first commit with all the autogenerated files, the rest will be manual changes.
I don't think anybody would want to review all the 60,747 files, that don't have any new code.

Commit:
Added all the autogenerated files.

ltratt commented:
Yes, I don't think I'll be reviewing this one as such :) It might be nice to point out a couple of interesting differences from the original Lua though, when it comes to "final" review.

Commit:
All of the files compile now.

All of the functions need do be made non static now.

Commit:
Every functions defined as extern

Commit:
Everything compiles now.

alex4o commented:
So the whole project now compiles and produces a final executable. The executable crashes with a segfault.

ltratt commented:
How far through does it get before crashing? Can you run e.g. hello world? Maybe paste some `gdb` output here?

alex4o commented:
No I can't get to the interpreter even. It is doing something strange with the stack the way I see it.  
![Screenshot from 2021-01-15 16-28-22](https://user-images.githubusercontent.com/3462438/104738979-dd7db200-574e-11eb-9faa-32a3d0a1ec0f.png)

ltratt commented:
Well, I suppose this shows that the security properties are working 👍:p

I did notice the "source file is newer than executable" warning in gdb, which sounds a bit fishy?

alex4o commented:
I found the current problem, there exist 2 `docall` functions in the lua code base. They accept different arguments, and the wrong one is being called. I am not perfectly sure how that can be fixed but I will look into it.

Commit:
Changed some of the functions to be non-preprocessed.

alex4o commented:
It works now!
![Screenshot from 2021-01-15 17-49-40](https://user-images.githubusercontent.com/3462438/104748188-1a9b7180-575a-11eb-806b-cb1c78d764e6.png)

ltratt commented:
Cool! Does the test suite as a whole run?

alex4o commented:
Some do, but the way lua's test are done, if one fails the other don't run, so I don't know how many pass or fail.

ltratt commented:
So my suggestion is: let's start looking at failures and see if we can fix them? The fact that you can already run the basic interpreter is a good sign, as it means, at the very least, that most of the system runs correctly!

alex4o commented:
They definitely can be fixed. The test runner actually causes a segfault too, so this is where I think I should start.

ltratt commented:
Sounds like a plan!

Commit:
Passing all of the unit test.

By not running the ones that fail.
The interpreter (main.lua) and io (files.lua) test are skipted because
they don't pass but that shouldn't be a problem. This commit also joins
some files together because of static functions.

Commit:
Solves all the problems with the files.lua test.

Again some of the preprocessor definitions were not done properly.

alex4o commented:
Some things of note IMO are: 
static variables for which the sizeof function is used need to be in the same file as the sizeof usage.
https://github.com/capablevms/lua-func/blob/9686752991a4658a0e3aa972f0c638387b4013c0/luaopen_io.c#L47-L77 

Actually I am not sure If I made this clear enough, but in this situation more functions are exported than normally, and now nearly every function is exported which in fact reduces the security, but this allows for grouping of functions in a very easy way just by deciding which object files will go in the final shared library.

ltratt commented:
>  Passing all of the unit test.
...
> By not running the ones that fail.

:p

ltratt commented:
Can you give a rough idea of how many tests N are failing out of M total?

alex4o commented:
The lua tests are very bad, and they aren't organized in any meaningful way. With the last changes, there is a single line in the:
files.lua file which is commented out, and the main.lua file is ignored, because to me it looks like it requires libreadline (couldn't understand why) to pass and that library doesn't work on CHERI. So lets say 15% of the tests fail, but only in the interpreter.

ltratt commented:
OK, let's edit the (or add a?!) top-level README just briefly documenting what's been commented out and roughly how many tests aren't running, just so that we don't forget. Once we've done that, I think we can merge this.

Commit:
Explain the project purpose in the README file.

alex4o commented:
This should be ready for merging.

ltratt commented on README.md:
```diff
@@ -1 +1,7 @@
\ No newline at end of file
+# lua-func
+
+The lua interpreter split in a way such that each function is in it's own file.
```
Every function is in a unique file? I thought some files contained more than one function?

[Also "Lua" with a capital "L" to be polite :)]

alex4o replied:
Done, and yes there are, but I can't seem to find them now

ltratt replied:
Well, if it's not true, we can't put it in the README :p Can you check and let me know if this statement is correct or not?

alex4o replied:
So this is why I took so long to respond to you because I was looking for them all morning. But I just remembered the name of the function that caused this problem and now I now which files are like that.

ltratt replied:
Is it just one pair (or more than 2?) functions or ... ? Should we say "all functions are split into single files except X and Y"? Also, can we spell out *why* we've put them in files? Maybe saying something like "This allows us to make each file a shared object and use capabilities to do A, B, and C."?

ltratt replied:
Your change looks good, but can we have a think about:

> Also, can we spell out why we've put them in files? Maybe saying something like "This allows us to make each file a shared object and use capabilities to do A, B, and C."?

cristianurlea replied:
Instructions on how to build/run/crash this would also be very welcome in the README.

alex4o replied:
The main idea is, not to directly write any capability code and still obtain results. Instructions on building and running should be added yes.

ltratt replied:
OK, let's add some stuff in along those lines. It doesn't have to be super extensive -- a moderate paragraph should be enough by the sounds of things.

alex4o replied:
I have a build script but I am not sure if I should commit it or not: https://gist.github.com/alex4o/2a72535101f078f4b390e8fb1487f5bb

cristianurlea replied:
I would suggest that you just inline the build script inside the README.md as a code block. It would be difficult to notice the build script amongst the sea of .c files anyway.

alex4o replied:
Is the README fine now?

ltratt replied:
Why wouldn't we include the build script in the repo directly? Maybe even converting it to a Makefile?

cristianurlea replied:
For clarity what i meant was: "I would suggest, at very minimum inlining the build script". I have no objections to the build script or an equivalent Makefile being in the repo directly.

Given the stated description for the repo, I would imagine this build script is what would allow us to select amongst the "different groupings of functions", eventually.

alex4o replied:
I can use the shell script to generate a Makefile, but that would reduce the understanding of what is actually happaning.

ltratt commented on README.md:
```diff
@@ -1 +1,7 @@
+
+However not all of the tests pass.
+
+One of the popen tests in files.lua:747 is comented out, and the entire main.lua file is ignored. This amounts to roughly 15% of the tests failing.
```
Line needs wrapping.

alex4o replied:
Do we need the text wrapping for markdown files, which will be mostly viewed in the browser which provides better line wrapping?

ltratt replied:
Indeed we do need to wrap -- long lines work very badly with `diff`.

alex4o commented:
Fixed the long lines.

alex4o commented:
Is this better. The commit messages aren't perfect but I would like not to squash the commits (except the readme ones) so the preserve the history.

ltratt commented on README.md:
```diff
@@ -1,9 +1,10 @@
 
 The Lua interpreter split in a way such that each function is in it's own file. Except for a few
-functions that were very hard to split up, for example the `lundump.c` file. 
+functions that were very hard to split up, for example the `lundump.c` file. Doing that would not
```
What does "doing that would not give additional CHERI security guarantees" mean?

alex4o replied:
Is it more understandable now?

ltratt commented on README.md:
```diff
@@ -1,9 +1,10 @@
-functions that were very hard to split up, for example the `lundump.c` file. Doing that would not
-give additional CHERI security guarantees. However this would allow for experimenting with 
-different groupings of functions. The main purpose of doing this is to get additional security
+functions that were very hard to split up, for example the `lundump.c` file. Just moving the 
```
The "just moving" sentence doesn't really make sense -- as, at least as I understand things, "just moving" is really what this PR is about, right? :) So I'm not sure what point's being made here. It's probably worth us getting this right, so I'm happy to bounce ideas back and forth.

alex4o replied:
Yes, but I am trying to articulate that this isn't enough, there needs to be some other work that is not in the repo, to obtain the final result.

alex4o replied:
@ltratt, any thoughts?

ltratt replied:
Can you propose some new text (even if it's rough)?

alex4o replied:
I think this explains it better, but I am not sure.

ltratt commented on README.md:
s/would/will/

ltratt replied:
Ping.

ltratt commented on README.md:
I think the "." after "out" should be a ":".

ltratt replied:
Ping.

Commit:
Add the two files which are copied directly from the Lua repository.

These are the lua_docall and ldo_call which are different, and hard to
split. Also the lundump file which has only one public function.

Commit:
Add the build script as a buildbot configuration.

Commit:
The file functions have incorrect definitions.

The exports were wrong, and didn't work work when moved to a different linux distribution.

ltratt commented:
@alex4o where are with this PR?

alex4o commented:
If you like what you see we can merge it. It builds on x86 and my rpi. I will reorder and reword the commits.

ltratt commented:
Do you want to do any squashing?

ltratt commented:
@alex4o ping.

Commit:
Add a main file which also includes utilities for time measuriment.

alex4o commented:
I force pushed. Can you review the commit messages, and if they make sense?

ltratt commented on main-time.c:
```diff
@@ -0,0 +1,97 @@
+
+  timespec_diff(&start_time, &end_time, &time_diff);
+
+  char * name = "";
```
Can we be consistent with `*` and spacing: we have `char * name` and a few lines later `FILE *res`. The common convention in C is `char *name` and personally I'd prefer that (it reduces one common source of C error).

ltratt commented:
The commit messages look fine, but I *think* there are unresolved comments from before?

alex4o commented:
I think my last commit should address all of the comments.

ltratt commented on README.md:
s/One/one/ (no capital letters after `:` :) )

alex4o replied:
I didn't know about that.

alex4o commented:
Can I squash now?

ltratt commented:
Please squash.

alex4o commented:
I kind of realized that buildbot wouldn't be able to build this, so it should just merge it normally.
